dotCMS/src/main/java/com/dotcms/keyvalue/business/KeyValueAPIImpl.java

```diff
@@ -0,0 +1,146 @@
+import com.liferay.portal.model.User;
+import com.liferay.util.StringPool;
+
+/**
+ * Implementation class for the {@link KeyValueAPI}.
+ * 
+ * @author Jose Castro
+ * @version 4.2.0
+ * @since Jun 19, 2017
+ *
+ */
+public class KeyValueAPIImpl implements KeyValueAPI {
+
+    protected final ContentletAPI contentletAPI;
+    protected final FolderAPI folderAPI;
+    protected final UserAPI userAPI;
+    protected final KeyValueCache cache;
+
+    /**
+     * Creates a new instance of the {@link KeyValueAPI}.
+     */
+    public KeyValueAPIImpl() {
+        this.contentletAPI = APILocator.getContentletAPI();
+        this.folderAPI = APILocator.getFolderAPI();
+        this.userAPI = APILocator.getUserAPI();
+        this.cache = CacheLocator.getKeyValueCache();
+    }
+
+    @Override
+    public List<KeyValue> get(final String key, final User user, final boolean respectFrontendRoles) {
+        List<KeyValue> results = this.cache.get(key);
+        if (null != results) {
+            return results;
+        }
+        results = queryKeyValues(key, -1, null, user, respectFrontendRoles);
+        if (null != results && !results.isEmpty()) {
+            this.cache.add(key, results);
+        }
+        return results;
+    }
+
+    @Override
+    public List<KeyValue> get(final String key, final long languageId, final User user, final boolean respectFrontendRoles) {
+        List<KeyValue> results = this.cache.getByLanguage(key, languageId);
+        if (null != results) {
+            return results;
+        }
+        results = queryKeyValues(key, languageId, null, user, respectFrontendRoles);
+        if (null != results && !results.isEmpty()) {
+            this.cache.addByLanguage(key, languageId, results);
+        }
+        return results;
+    }
+
+    @Override
+    public List<KeyValue> get(final String key, final ContentType contentType, final User user,
+                    final boolean respectFrontendRoles) {
+        List<KeyValue> results = this.cache.getByContentType(key, contentType.id());
+        if (null != results) {
+            return results;
+        }
+        results = queryKeyValues(key, -1, contentType, user, respectFrontendRoles);
+        if (null != results && !results.isEmpty()) {
+            this.cache.addByContentType(key, contentType.id(), results);
+        }
+        return results;
+    }
+
+    @Override
+    public KeyValue get(final String key, final long languageId, final ContentType contentType, final User user,
+                    final boolean respectFrontendRoles) {
+        if (languageId > -1 && null != contentType && UtilMethods.isSet(contentType.id())) {
+            final KeyValue keyValue = this.cache.getByLanguageAndContentType(key, languageId, contentType.id());
+            if (null != keyValue) {
+                return keyValue;
+            }
+        }
+
+        final List<KeyValue> result = queryKeyValues(key, languageId, contentType, user, respectFrontendRoles);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') "NullPointerException" will be thrown when invoking method "queryKeyValues()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)